### ISSUE TITLE
Missed optimization in Euclid Remainder

### ISSUE THREAD

**Issue by zommiommy** [llvm/llvm-project#64305]
Hi, @vigna, @progval, and I, noticed a missed optimization in the computation of Euclid remainder:
```C
int div_euclid(int num) {
    int q = num / 8;
    if (num % 8 < 0)
        return 8 > 0 ? q - 1 : q + 1;
    return q;
}

int rem_euclid(int num) {
    int r = num % 8;
    if (r < 0) {
        return r + (8 < 0 ? -8 : 8);
    }
    return r;
}
```
https://godbolt.org/z/fKr1xsWYT

while div_euclid is correctly optimized to:
```asm
mov     eax, edi
sar     eax, 3
ret
```

while rem_euclid is correctly optimized to:
```asm            
lea     eax, [rdi + 7]
test    edi, edi
cmovns  eax, edi
and     eax, -8
mov     ecx, edi
sub     ecx, eax
neg     eax
test    ecx, ecx
lea     eax, [rdi + rax + 8]
cmovns  eax, ecx
ret
```
instead of:
```asm
mov     eax, edi
and     eax, 7
ret
```

**Comment by antoniofrighetto** [comment]
Thanks for reporting! Quite likely InstCombine is not catching it, proof: https://alive2.llvm.org/ce/z/ma6se5.

**Comment by antoniofrighetto** [comment]
Closing this as landed with 211692137af45fcf6d6b14582e004a746f9e5b2e.
